Node revamp: Routing (#27796)
TheOdinProject/curriculum: nodeJS/express/routing.md

### Introduction

In the previous lesson, we broke down a basic Express app and a request's journey through it. However, in a real app, we'd want to handle many different types of requests for different things.

Routes essentially just match a request's HTTP verb (e.g. GET or POST) and URL path to the appropriate set of middleware functions - the controllers. We will explore controllers and middleware in more detail in the next lesson. For now, let's dive into how we can use routes.

### Lesson overview

This section contains a general overview of topics that you will learn in this lesson.

> - Describe how routes are defined.
> - Match routes to specific HTTP verbs and paths.
> - Explain route parameters and query parameters.
> - Extract routes to their own routers.

Reviewer: The second lesson overview can be skipped altogether as the first one encompasses it. Also, reworded the router overview (suggestions welcome):
```suggestion
- Describe how routes are defined.
- Explain route parameters and query parameters.
- Setup new routers on a path
```

### The anatomy of a route

Let's return to our sole route in our previous basic Express app.

```javascript
app.get("/", (req, res) => res.send("Hello, world!"));
```

> `app.get("/" ...` tells us that this route will match any requests that go through `app` (our whole server) that have the GET HTTP verb and are for the `/` path. If instead we had the following:

Reviewer: Wording suggestion:
```suggestion
`app.get("/" ...` tells us that this route will match any `GET` requests that go through `app` (our whole server) router for the `/` path. If instead we had the following:
```

```javascript
app.post("/messages", (req, res) => res.send("This is where you can see any messages."));
```

That would tell us the route matches any POST requests to the `/messages` path of our `app`. If you sent a GET request to the `/messages` path, it would not match this route. Each HTTP verb has its own Express route method, and you can also use [app.all()](https://expressjs.com/en/api.html#app.all) to make a route match all verbs.

<div class="lesson-note" markdown="1">

#### HTTP verbs

> [There are many HTTP verbs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods), but for now, we will primarily be using GET (used to retrieve data from the server only) and POST (used to send data to the server e.g. forms). Later, when we cover REST APIs, you will often also encounter verbs like PUT and DELETE.

Reviewer: Not necessarily true. Learners will likely need to use `PUT` and `DELETE` in the inventory application (which is before the REST APIs section)

Reviewer: I missed that. Thanks for clarifying.

### Routers

So far, we've not been using many routes, and all routes we've shown have been attached to `app`, our server itself. In a real application with lots of routes, we'd probably want to organize our routes into groups and extract each group out to their own file. We could also then more easily write things that affect only the routes in that file, and not any others.

Say we were making a library app and we wanted pages that dealt with books and pages that dealt with authors. That's on top of the homepage and any other miscellaneous pages like "about" or "contact".

We might want our server to handle the following routes:

```text
GET /
GET /about
GET /contact
POST /contact

GET /books
GET /books/:bookId
GET /books/:bookId/reserve
POST /books/:bookId/reserve

GET /authors
GET /authors/:authorId
```

> It'd be nice if we could extract the route groups to their own files. It'd also be nice if the server didn't need to check any routes that didn't start with `/authors` if a request was made to a path that started with `/authors`. We can do that using routers!

Reviewer: > It'd also be nice if the server didn't need to check any routes that didn't start with `/authors` if a request was made to a path that started with `/authors`

^ this sentence is confusing, can you explain what it's trying to say?

Author: The original intent is that if you have 30 routes in a `/books` router and a request is for `/authors/*`, it'll straight up skip the routes in the `/books` router.

But thinking about it again, I'm not really sure it's an "issue" worth mentioning. Should be perfectly sufficient to say that a request is passed into the matching router, and that's it.

Reviewer: Hmm, not sure if it matters to the learners how many routes the server has to match before it finds the correct one.

Author: Makes sense in hindsight - this can be omitted
